### PR TITLE
Enable multiple fonts and paragraph alignment in coretext

### LINKED ISSUE #1013: Support setting fonts and paragraph styling in DWrtiteWrapper TextLayout
When moving to DWrite we regressed this functionality and today we do not support setting font for different attributed string ranges or support paragraph styling

### PR DESCRIPTION
**Pull request by aballway**
Fixes regression which defaulted to left-aligned text, and adds support for selecting different fonts and changing writing direction

Fixes #1013

### CODE REVIEW

**Comment by rajsesh-msft** [comment]
>     //  These can be done using the DWrite IDWriteTextFormat range property methods.

Update TODO to correctly reflect what is remaining?

---
Refers to: Frameworks/CoreText/DWriteWrapper.mm:287 in 97feed1. [](commit_id = 97feed1315c992d066164bd3737259cbb38ef092, deletion_comment = False)

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 1]
```diff
@@ -232,12 +290,25 @@ static CFStringRef _CFStringFromLocalizedString(IDWriteLocalizedStrings* localiz
...
+
+        CTFontRef font = static_cast<CTFontRef>([attribs objectForKey:static_cast<NSString*>(kCTFontAttributeName)]);
         if (font != nil) {
-            fontSize = [font pointSize];
+            CGFloat fontSize = CTFontGetSize(font);
+            DWRITE_FONT_WEIGHT weight = DWRITE_FONT_WEIGHT_REGULAR;
+            DWRITE_FONT_STYLE style = DWRITE_FONT_STYLE_NORMAL;
+            DWRITE_FONT_STRETCH stretch = DWRITE_FONT_STRETCH_NORMAL;
+            __InitDWriteFontPropertiesFromName(CTFontCopyName(font, kCTFontFullNameKey), &weight, &stretch, &style);
+            CFStringRef name = _DWriteGetFamilyNameForFontName(CTFontCopyName(font, kCTFontFullNameKey));
```
Comment (rajsesh-msft) [inline Frameworks/CoreText/DWriteWrapper.mm]: >CTFontCopyName(font, kCTFontFullNameKey) [](start = 63, length = 40)

nit: cache this result instead of calling it twice?

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 2]
```diff
@@ -232,12 +290,25 @@ static CFStringRef _CFStringFromLocalizedString(IDWriteLocalizedStrings* localiz
...
+            std::vector<wchar_t> fontName(CFStringGetLength(name) + 1);
+            CFStringGetCharacters(name, CFRangeMake(0, fontName.size()), reinterpret_cast<UniChar*>(fontName.data()));
+
+            THROW_IF_FAILED(textLayout->SetFontSize(fontSize, DWRITE_TEXT_RANGE{ attributeRange.location, attributeRange.length }));
```
Comment (rajsesh-msft) [inline Frameworks/CoreText/DWriteWrapper.mm]: >DWRITE_TEXT_RANGE{ attributeRange.location, attributeRange.length } [](start = 62, length = 67)

const DWRITE_TEXT_RANGE range {...} and then use them.  compiler will likely optimize it the same, but the code would be cleaner.

File: tests/unittests/CoreText/CTLineTests.mm [hunk 3]
```diff
@@ -166,24 +166,23 @@
     CFRelease(startTruncated);
 
     CTLineRef endTruncated = CTLineCreateTruncatedLine(line, 40.0, kCTLineTruncationEnd, NULL);
-    EXPECT_EQ(3, CTLineGetGlyphCount(endTruncated));
+    EXPECT_EQ(2, CTLineGetGlyphCount(endTruncated));
 
-    // Truncated should be ABC in one run
+    // Truncated should be AB in one run
```
Comment (rajsesh-msft) [inline tests/unittests/CoreText/CTLineTests.mm]: why are these changing? seems like a regression?

File: tests/testapps/CTCatalog/CTCatalog/Tests/CTCAlignmentTestViewController.m [hunk 4]
```diff
@@ -181,13 +182,24 @@ - (void)drawTests {
         [[_textField.text stringByReplacingOccurrencesOfString:@"\\n" withString:@"\n"] stringByReplacingOccurrencesOfString:@"\\t"
                                                                                                                   withString:@"\t"];
     [self.view addSubview:_rightView];
+
+    // Create justified text
+    _justifiedView = [[CTAlignmentTestView alloc] initWithFrame:CGRectMake(20, 620, 400, 200)];
```
Comment (rajsesh-msft) [inline tests/testapps/CTCatalog/CTCatalog/Tests/CTCAlignmentTestViewController.m]: Sweet!

File: Frameworks/UIKit/UIFont.mm [hunk 5]
```diff
@@ -141,8 +141,7 @@ + (UIFont*)systemFontOfSize:(CGFloat)size weight:(CGFloat)weight {
 + (UIFont*)boldSystemFontOfSize:(CGFloat)size {
     // TODO 5785385: Using clumsy fontWithDescriptor to initialize here, so that _descriptor is initialized
     // Clean this up a bit once fontDescriptor gets better support
-    UIFontDescriptor* fontDes =
-        [[UIFontDescriptor fontDescriptorWithName:@"Arial Bold" size:size] fontDescriptorWithSymbolicTraits:UIFontDescriptorTraitBold];
```
Comment (rajsesh-msft) [inline Frameworks/UIKit/UIFont.mm]: is this because fontDescriptorWithSymbolicTraits is broken?

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 6]
```diff
@@ -36,6 +36,11 @@
 static const wchar_t* c_defaultUserLanguage = L"en-us";
 static const wchar_t* c_defaultFontName = L"Segoe UI";
 
+static void __InitDWriteFontPropertiesFromName(CFStringRef fontName,
```
Comment (rajsesh-msft) [inline Frameworks/CoreText/DWriteWrapper.mm]: >static [](start = 0, length = 6)

You dont need static in the forward declaration.

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 7]
```diff
@@ -173,21 +178,74 @@ static CFStringRef _CFStringFromLocalizedString(IDWriteLocalizedStrings* localiz
...
+    DWRITE_FONT_STYLE style = DWRITE_FONT_STYLE_NORMAL;
+    DWRITE_FONT_STRETCH stretch = DWRITE_FONT_STRETCH_NORMAL;
     if (font != nil) {
-        fontSize = [font pointSize];
+        fontSize = CTFontGetSize(font);
+        __InitDWriteFontPropertiesFromName(CTFontCopyName(font, kCTFontFullNameKey), &weight, &stretch, &style);
+        CFStringRef name = _DWriteGetFamilyNameForFontName(CTFontCopyName(font, kCTFontFullNameKey));
+        fontName.resize(CFStringGetLength(name) + 1);
+        CFStringGetCharacters(name, CFRangeMake(0, fontName.size()), reinterpret_cast<UniChar*>(fontName.data()));
     }
```
Comment (rajsesh-msft) [inline Frameworks/CoreText/DWriteWrapper.mm]: Almost seems like a need for _DWriteGetFontFamilyNameAndProperties(...), so all of this can be done in one shot in multiple functions.  Right now, this segment is repeated.

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 8]
```diff
@@ -173,21 +178,74 @@ static CFStringRef _CFStringFromLocalizedString(IDWriteLocalizedStrings* localiz
...
+                case kCTCenterTextAlignment:
+                    dwriteAlignment = DWRITE_TEXT_ALIGNMENT_CENTER;
+                    break;
+                case kCTJustifiedTextAlignment:
+                    dwriteAlignment = DWRITE_TEXT_ALIGNMENT_JUSTIFIED;
+                    break;
+                case kCTLeftTextAlignment:
+                case kCTNaturalTextAlignment:
+                default:
+                    dwriteAlignment = DWRITE_TEXT_ALIGNMENT_LEADING;
```
Comment (rajsesh-msft) [inline Frameworks/CoreText/DWriteWrapper.mm]: >dwriteAlignment = DWRITE_TEXT_ALIGNMEN [](start = 20, length = 38)

just break is fine here.

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 9]
```diff
@@ -173,21 +178,74 @@ static CFStringRef _CFStringFromLocalizedString(IDWriteLocalizedStrings* localiz
...
+            }
+
+            THROW_IF_FAILED(textFormat->SetTextAlignment(dwriteAlignment));
+        }
+
+        CTWritingDirection direction;
+        if (CTParagraphStyleGetValueForSpecifier(settings, kCTParagraphStyleSpecifierBaseWritingDirection, sizeof(direction), &direction)) {
+            DWRITE_READING_DIRECTION dwriteDirection = DWRITE_READING_DIRECTION_LEFT_TO_RIGHT;
+            switch (direction) {
+                case kCTWritingDirectionRightToLeft:
```
Comment (rajsesh-msft) [inline Frameworks/CoreText/DWriteWrapper.mm]: we need to add a test in the test app for the RTL text, given we are adding this now.

**Comment by rajsesh-msft** [comment]
:clock1:

File: Frameworks/UIKit/UIFont.mm (hunk 5 above)
Comment (aballway) [inline Frameworks/UIKit/UIFont.mm]: Right now it's just stubbed out

File: Frameworks/CoreText/DWriteWrapper.mm (hunk 9 above)
Comment (aballway) [inline Frameworks/CoreText/DWriteWrapper.mm]: This is tested in the ParagraphStyle Tests

File: Frameworks/CoreText/DWriteWrapper.mm (hunk 1 above)
Comment (ms-jihua) [inline Frameworks/CoreText/DWriteWrapper.mm]: >name [](start = 24, length = 4)

can we name this 'familyName' to avoid ambiguity?

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 10]
```diff
@@ -173,21 +178,74 @@ static CFStringRef _CFStringFromLocalizedString(IDWriteLocalizedStrings* localiz
...
     CGFloat fontSize = kCTFontSystemFontSize;
+    std::vector<wchar_t> fontName(c_defaultFontName, c_defaultFontName + wcslen(c_defaultFontName));
+    DWRITE_FONT_WEIGHT weight = DWRITE_FONT_WEIGHT_REGULAR;
+    DWRITE_FONT_STYLE style = DWRITE_FONT_STYLE_NORMAL;
+    DWRITE_FONT_STRETCH stretch = DWRITE_FONT_STRETCH_NORMAL;
     if (font != nil) {
-        fontSize = [font pointSize];
+        fontSize = CTFontGetSize(font);
+        __InitDWriteFontPropertiesFromName(CTFontCopyName(font, kCTFontFullNameKey), &weight, &stretch, &style);
+        CFStringRef name = _DWriteGetFamilyNameForFontName(CTFontCopyName(font, kCTFontFullNameKey));
```
Comment (ms-jihua) [inline Frameworks/CoreText/DWriteWrapper.mm]: >name [](start = 20, length = 4)

can we name this 'familyName' to avoid ambiguity?

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 11]
```diff
@@ -173,21 +178,74 @@ static CFStringRef _CFStringFromLocalizedString(IDWriteLocalizedStrings* localiz
...
+        fontName.resize(CFStringGetLength(name) + 1);
+        CFStringGetCharacters(name, CFRangeMake(0, fontName.size()), reinterpret_cast<UniChar*>(fontName.data()));
```
Comment (ms-jihua) [inline Frameworks/CoreText/DWriteWrapper.mm]: >fontName [](start = 96, length = 8)

actually wait... you just put the family name into the font name char buffer... what's going on?

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 12]
```diff
@@ -232,12 +290,25 @@ static CFStringRef _CFStringFromLocalizedString(IDWriteLocalizedStrings* localiz
...
         if (font != nil) {
-            fontSize = [font pointSize];
+            CGFloat fontSize = CTFontGetSize(font);
+            DWRITE_FONT_WEIGHT weight = DWRITE_FONT_WEIGHT_REGULAR;
+            DWRITE_FONT_STYLE style = DWRITE_FONT_STYLE_NORMAL;
+            DWRITE_FONT_STRETCH stretch = DWRITE_FONT_STRETCH_NORMAL;
+            __InitDWriteFontPropertiesFromName(CTFontCopyName(font, kCTFontFullNameKey), &weight, &stretch, &style);
+            CFStringRef name = _DWriteGetFamilyNameForFontName(CTFontCopyName(font, kCTFontFullNameKey));
+            std::vector<wchar_t> fontName(CFStringGetLength(name) + 1);
+            CFStringGetCharacters(name, CFRangeMake(0, fontName.size()), reinterpret_cast<UniChar*>(fontName.data()));
```
Comment (ms-jihua) [inline Frameworks/CoreText/DWriteWrapper.mm]: >fontName [](start = 100, length = 8)

same question as above

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 13]
```diff
@@ -173,21 +178,74 @@ static CFStringRef _CFStringFromLocalizedString(IDWriteLocalizedStrings* localiz
...
+    THROW_IF_FAILED(
+        dwriteFactory->CreateTextFormat(fontName.data(), NULL, weight, style, stretch, fontSize, c_defaultUserLanguage, &textFormat));
+
+    CTParagraphStyleRef settings =
+        static_cast<CTParagraphStyleRef>([attribs valueForKey:static_cast<NSString*>(kCTParagraphStyleAttributeName)]);
+    if (settings) {
+        CTTextAlignment alignment = kCTNaturalTextAlignment;
+        if (CTParagraphStyleGetValueForSpecifier(settings, kCTParagraphStyleSpecifierAlignment, sizeof(CTTextAlignment), &alignment)) {
+            DWRITE_TEXT_ALIGNMENT dwriteAlignment = DWRITE_TEXT_ALIGNMENT_LEADING;
+            switch (alignment) {
```
Comment (ms-jihua) [inline Frameworks/CoreText/DWriteWrapper.mm]: consider moving this switch statement out into its own helper:
- you won't have to break since you can just early return
- makes it easy to reuse
- breaks the code into smaller chunks

File: tests/unittests/CoreText/CTFrameTests.mm [hunk 14]
```diff
@@ -25,7 +25,7 @@
 static constexpr float c_arbitraryFloat = 314159.2717;
 
 static NSAttributedString* getAttributedString(NSString* str) {
-    UIFontDescriptor* fontDescriptor = [UIFontDescriptor fontDescriptorWithName:@"Times New Roman" size:40];
+    UIFontDescriptor* fontDescriptor = [UIFontDescriptor fontDescriptorWithName:@"Segoe UI" size:40];
```
Comment (ms-jihua) [inline tests/unittests/CoreText/CTFrameTests.mm]: >Segoe UI [](start = 82, length = 8)

Any particular context for changing this?

**Comment by ms-jihua** [comment]
:clock1:

File: Frameworks/CoreText/DWriteWrapper.mm (hunk 11 above)
Comment (aballway) [inline Frameworks/CoreText/DWriteWrapper.mm]: CreateTextFormat expects the font family name.  I'm gonna rename all of these variables to make it clearer what is what

---
In reply to: [79929428](https://github.com/Microsoft/WinObjC/pull/1021#discussion_r79929428) [](ancestors = 79929428)

File: tests/unittests/CoreText/CTFrameTests.mm (hunk 14 above)
Comment (aballway) [inline tests/unittests/CoreText/CTFrameTests.mm]: The test values were done using Segoe, so when we could actually use Times it threw off all of the numbers.  tl;dr it's a lazier fix than adjusting all of the values for a different font unrelated to the test

---
In reply to: [79930359](https://github.com/Microsoft/WinObjC/pull/1021#discussion_r79930359) [](ancestors = 79930359)

File: tests/unittests/CoreText/CTLineTests.mm (hunk 3 above)
Comment (aballway) [inline tests/unittests/CoreText/CTLineTests.mm]: UIFont seems to use Arial as default, so these are the values actually using that font.

---
In reply to: [79909850](https://github.com/Microsoft/WinObjC/pull/1021#discussion_r79909850) [](ancestors = 79909850)

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 15]
```diff
@@ -756,7 +822,7 @@ HRESULT _DWriteCreateFontFaceWithName(CFStringRef name, IDWriteFontFace** outFon
     DWRITE_FONT_STRETCH stretch = DWRITE_FONT_STRETCH_NORMAL;
     DWRITE_FONT_STYLE style = DWRITE_FONT_STYLE_NORMAL;
 
-    __InitDWriteFontPropertiesFromName(name, &weight, &stretch, &style);
+    __InitDWriteFontPropertiesFromName(name, &weight, &stretch, &style, nullptr);
 
     CFStringRef familyName = _DWriteGetFamilyNameForFontName(name);
```
Comment (ms-jihua) [inline Frameworks/CoreText/DWriteWrapper.mm]: uhhh... just pass it something from the extended helper lol

Comment (aballway) [inline Frameworks/CoreText/DWriteWrapper.mm]: whoops, didn't catch that

---
In reply to: [79946189](https://github.com/Microsoft/WinObjC/pull/1021#discussion_r79946189) [](ancestors = 79946189)

**Comment by mnithish** [comment]
>     // Get font family name details so that can be used here. For now this is hardcoded to "Gabriola".

Sorry I think I left this TODO out. Can you delete it for me please?

---
Refers to: Frameworks/CoreText/DWriteWrapper.mm:187 in cd93a98. [](commit_id = cd93a988446fad9cca334d8bc37c97ebd014891d, deletion_comment = False)

**Comment by mnithish** [comment]
>     // Per string range attribute handling will be done in _CreateDWriteTextLayout.

What about this feature? We again are only considering the attribute value set at the range.location. When we are at this change we should add support for this as-well

---
Refers to: Frameworks/CoreText/DWriteWrapper.mm:190 in cd93a98. [](commit_id = cd93a988446fad9cca334d8bc37c97ebd014891d, deletion_comment = False)

**Comment by mnithish** [comment]
>     // Get font family name details so that can be used here. For now this is hardcoded to "Gabriola".

Never mind I will handle it in my CR

---
In reply to: [248784512](https://github.com/Microsoft/WinObjC/pull/1021#issuecomment-248784512) [](ancestors = 248784512)

---
Refers to: Frameworks/CoreText/DWriteWrapper.mm:187 in cd93a98. [](commit_id = cd93a988446fad9cca334d8bc37c97ebd014891d, deletion_comment = False)

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 16]
```diff
@@ -173,21 +191,54 @@ static CFStringRef _CFStringFromLocalizedString(IDWriteLocalizedStrings* localiz
 
     // TODO: #1001 attribs can be nil here?
```
Comment (mnithish) [inline Frameworks/CoreText/DWriteWrapper.mm]: >// TODO: #1001 attribs can be nil here? [](start = 4, length = 39)

Is this TODO still valid?

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 17]
```diff
@@ -756,9 +818,9 @@ HRESULT _DWriteCreateFontFaceWithName(CFStringRef name, IDWriteFontFace** outFon
     DWRITE_FONT_STRETCH stretch = DWRITE_FONT_STRETCH_NORMAL;
     DWRITE_FONT_STYLE style = DWRITE_FONT_STYLE_NORMAL;
 
-    __InitDWriteFontPropertiesFromName(name, &weight, &stretch, &style);
+    CFStringRef familyName;
+    __InitDWriteFontPropertiesFromName(name, &weight, &stretch, &style, &familyName);
```
Comment (mnithish) [inline Frameworks/CoreText/DWriteWrapper.mm]: Who owns the familyName memory?

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 18]
```diff
@@ -618,19 +682,17 @@ CFStringRef _DWriteGetFamilyNameForFontName(CFStringRef fontName) {
...
+static void __InitDWriteFontPropertiesFromName(
+    CFStringRef fontName, DWRITE_FONT_WEIGHT* weight, DWRITE_FONT_STRETCH* stretch, DWRITE_FONT_STYLE* style, CFStringRef* familyName) {
     // Set some defaults for when weight/stretch/style are not mentioned in the name
     // Since this is a file-private helper, assume the pointers are safe to dereference.
     *weight = DWRITE_FONT_WEIGHT_NORMAL;
     *stretch = DWRITE_FONT_STRETCH_NORMAL;
     *style = DWRITE_FONT_STYLE_NORMAL;
 
-    CFStringRef familyName = _DWriteGetFamilyNameForFontName(fontName);
+    *familyName = _DWriteGetFamilyNameForFontName(fontName);
```
Comment (mnithish) [inline Frameworks/CoreText/DWriteWrapper.mm]: >_DWriteGetFamilyNameForFontName [](start = 18, length = 31)

The string that is being returned here is not retained, how will the memory management work here?

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 19]
```diff
@@ -224,20 +275,33 @@ static CFStringRef _CFStringFromLocalizedString(IDWriteLocalizedStrings* localiz
...
     //  - set indentation
-    //  - set font characteristics
+    //  - set kern
     //  - etc.
     //  These can be done using the DWrite IDWriteTextFormat range property methods.
     NSRange attributeRange;
     for (size_t i = 0; i < [ts->_attributedString length]; i += attributeRange.length) {
         NSDictionary* attribs = [ts->_attributedString attributesAtIndex:i
                                                    longestEffectiveRange:&attributeRange
                                                                  inRange:{ i, [ts->_attributedString length] - i }];
```
Comment (mnithish) [inline Frameworks/CoreText/DWriteWrapper.mm]: It will be nice to have a test for this in CTCatalog

**Comment by mnithish** [comment]
:clock1:

File: Frameworks/CoreText/DWriteWrapper.mm (hunk 18 above)
Comment (ms-jihua) [inline Frameworks/CoreText/DWriteWrapper.mm]: I mean, this currently works just fine because the string is being held in a dictionary and thus is guaranteed to last for as long as we need it, and has balanced retain count. Is your concern just that if we didn't know about that implementation detail, we wouldn't be able to guarantee that this'd be the case, because autorelease wouldn't do the correct thing in this situation?

---
In reply to: [79964653](https://github.com/Microsoft/WinObjC/pull/1021#discussion_r79964653) [](ancestors = 79964653)

Comment (mnithish) [inline Frameworks/CoreText/DWriteWrapper.mm]: Yes, we should not assume the underlying data structure and follow the CF memory model - a *Get* will return an autorelease object and it has be retained and freed appropriately 

---
In reply to: [79965724](https://github.com/Microsoft/WinObjC/pull/1021#discussion_r79965724) [](ancestors = 79965724,79964653)

File: Frameworks/CoreText/DWriteWrapper.mm (hunk 19 above)
Comment (aballway) [inline Frameworks/CoreText/DWriteWrapper.mm]: This is being tested by CTLine_CreateTruncatedLineTruncationToken which makes sure the ranges are broken properly

---
In reply to: [79964738](https://github.com/Microsoft/WinObjC/pull/1021#discussion_r79964738) [](ancestors = 79964738)

File: Frameworks/CoreText/DWriteWrapper.mm (hunk 17 above)
Comment (aballway) [inline Frameworks/CoreText/DWriteWrapper.mm]: It's owned by the dictionary which has all of the family names

---
In reply to: [79964582](https://github.com/Microsoft/WinObjC/pull/1021#discussion_r79964582) [](ancestors = 79964582)

File: Frameworks/CoreText/DWriteWrapper.mm (hunk 18 above)
Comment (aballway) [inline Frameworks/CoreText/DWriteWrapper.mm]: Fixing that path is outside of the scope of what I want this PR to be

---
In reply to: [79976593](https://github.com/Microsoft/WinObjC/pull/1021#discussion_r79976593) [](ancestors = 79976593,79965724,79964653)

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 20]
```diff
@@ -224,20 +275,33 @@ static CFStringRef _CFStringFromLocalizedString(IDWriteLocalizedStrings* localiz
...
-        CGFloat fontSize = kCTFontSystemFontSize;
-        UIFont* font = [attribs objectForKey:static_cast<NSString*>(kCTFontAttributeName)];
+
+        const DWRITE_TEXT_RANGE dwriteRange = { attributeRange.location, attributeRange.length };
```
Comment (rajsesh-msft) [inline Frameworks/CoreText/DWriteWrapper.mm]: >dwriteRange [](start = 32, length = 11)

Would still like this to be declared closer to where it is used than out here.
